ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Tables.hs

```diff
@@ -0,0 +1,551 @@
+-- rest of the in-memory Ledger State in scope.
+--
+-- The mk can be instantiated to anything that is map-like, i.e. that expects
+-- two type parameters, the key and the value. In particular, we provide here
+-- the usual structures that we expect to deal with, namely:
+--
+-- - @ValuesMK@: a map of keys to values
+--
+-- - @KeysMK@: a set of keys
+--
+-- - @DiffMK@: a map of keys to differences
+--
+-- - @TrackingMK@: the product of both @ValuesMK@ and @DiffMK@
+--
+-- - @SeqDiffMK@: an efficient sequence of differences
+--
+-- - @CodecMK@: encoder and decoder for the key and value
+--
+-- Using the @TableStuff@ class, we can manipulate the mk on the Ledger Tables,
+-- and we can extract Ledger Tables from a Ledger State which will share the
+-- same mk, or we can replace the Ledger Tables associated to a particular
+-- in-memory Ledger State.
+--
+-- The general idea is that we keep a @SeqDiffMK@ on the @DbChangelog@ and by
+-- reading values from the backing store and then forwarding them through the
+-- sequence of diffs, we obtain partial (or total but that would be pointlessly
+-- inefficient) views on the Ledger State as if the data was not split (see
+-- 'withBlockReadSets').
+--
+-- The @DbChangelog@ keeps also a sequence of @LedgerState blk EmptyMK@ which
+-- represent only the in-memory side. Then when applying a Ledger rule on top of
+-- a Ledger State, we:
+--
+-- 1. Take the appropriate in-memory Ledger State from the @DbChangelog@, (by
+--    means of 'ledgerDbPrefix').
+--
+-- 2. Consult the disk for the values needed by the rule (for example if
+--    applying a block, we only need the TxIns of the transactions in the block)
+--    (see 'getBlockKeySets' and 'readDb').
+--
+-- 3. Forward the values read from the disk through the @SeqDiffMK@ up until the
+--    desired Ledger State (see 'forwardValues').
+--
+-- 4. Associate the Ledger State from (1) and the forwarded tables of (3) and
+--    then call the Ledger rule.
+--
+-- 5. Get back a @DiffMK@ with differences to the provided values that will then
+--    have to be stored in the @DbChangelog@ (see the type of
+--    'applyBlockLedgerResult').
+module Ouroboros.Consensus.Ledger.Tables (
+    LedgerStateKind
+  , ShowLedgerState (..)
+    -- * UTxO HD
+    -- ** Isolating the tables
+  , InMemory (..)
+  , StowableLedgerTables (..)
+  , TableStuff (..)
+  , TickedTableStuff (..)
+  , mapOverLedgerTables
+  , mapOverLedgerTablesTicked
+  , overLedgerTables
+  , overLedgerTablesTicked
+  , zipOverLedgerTables
+  , zipOverLedgerTablesTicked
+    -- ** Tables values
+  , ApplyMapKind
+  , ApplyMapKind' (..)
+  , MapKind
+  , showsApplyMapKind
+    -- *** Mediators
+  , CodecMK (..)
+  , DiffMK
+  , EmptyMK
+  , IsMapKind (..)
+  , KeysMK
+  , NameMK (..)
+  , QueryMK
+  , SeqDiffMK
+  , TrackingMK
+  , ValuesMK
+    -- ** Serialization
+  , SufficientSerializationForAnyBackingStore (..)
+  , valuesMKDecoder
+  , valuesMKEncoder
```

Comment:
The grouping of imports does not match the section structure in this module, which is probably nice to have

Reply:
Done